warehouse/router/tracker.go

```diff
@@ -10,196 +10,170 @@ import (
+	nowSQL := r.getNowSQL()
+	timeWindow := r.calculateTimeWindow(warehouse)
+
 	query := fmt.Sprintf(`
-				SELECT
-				  created_at
-				FROM
-				  %[1]s
-				WHERE
-				  source_id = $1 AND
-				  destination_id = $2 AND
-				  created_at > %[2]s - $3 * INTERVAL '1 MIN' AND
-				  created_at < %[2]s - $4 * INTERVAL '1 MIN'
-				ORDER BY
-				  id DESC
-				LIMIT
-				  1;
-				`,
-		warehouseutils.WarehouseStagingFilesTable,
+		SELECT created_at
+		FROM `+whutils.WarehouseStagingFilesTable+`
+		WHERE source_id = $1
+		  AND destination_id = $2
+		  AND created_at > %[1]s - $3 * INTERVAL '1 MIN'
+		  AND created_at < %[1]s - $4 * INTERVAL '1 MIN'
+		ORDER BY id DESC
+		LIMIT 1;`,
 		nowSQL,
 	)
-	queryArgs := []interface{}{
-		source.ID,
-		destination.ID,
+	queryArgs := []any{
+		warehouse.Source.ID,
+		warehouse.Destination.ID,
 		2 * timeWindow / time.Minute,
 		timeWindow / time.Minute,
 	}
 
+	var createdAt sql.NullTime
 	err := r.db.QueryRowContext(ctx, query, queryArgs...).Scan(&createdAt)
-	if errors.Is(err, sql.ErrNoRows) {
-		return nil
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
+		return time.Time{}, nil
 	}
 	if err != nil {
-		return fmt.Errorf("fetching last upload time for source: %s and destination: %s: %w", source.ID, destination.ID, err)
+		return time.Time{}, fmt.Errorf("fetching oldest staging file for source %s and destination %s: %w",
+			warehouse.Source.ID, warehouse.Destination.ID, err)
 	}
-
 	if !createdAt.Valid {
-		return fmt.Errorf("invalid last upload time for source: %s and destination: %s", source.ID, destination.ID)
+		return time.Time{}, fmt.Errorf("invalid created_at time for source %s and destination %s",
+			warehouse.Source.ID, warehouse.Destination.ID)
 	}
+	return createdAt.Time, nil
+}
 
-	query = `
-				SELECT
-				  EXISTS (
-					SELECT
-					  1
-					FROM
-					  ` + warehouseutils.WarehouseUploadsTable + `
-					WHERE
-					  source_id = $1 AND
-					  destination_id = $2 AND
-					  (
-						status = $3
-						OR status = $4
-						OR status LIKE $5
-					  ) AND
-					  updated_at > $6
-				  );
-	`
-	queryArgs = []interface{}{
-		source.ID,
-		destination.ID,
+func (r *Router) calculateTimeWindow(warehouse *model.Warehouse) time.Duration {
+	timeWindow := r.config.uploadBufferTimeInMin.Load()
+	syncFrequency := warehouse.GetStringDestinationConfig(r.conf, model.SyncFrequencySetting)
+	if syncFrequency != "" {
+		if value, err := strconv.Atoi(syncFrequency); err == nil {
+			timeWindow += time.Duration(value) * time.Minute
+		}
+	}
+	return timeWindow
+}
+
+func (r *Router) checkUploadStatus(ctx context.Context, warehouse *model.Warehouse, createdAt time.Time) (bool, error) {
+	query := `
+		SELECT EXISTS (
+			SELECT 1
+			FROM ` + whutils.WarehouseUploadsTable + `
+			WHERE source_id = $1 AND destination_id = $2 AND
+				  (status = $3 OR status = $4 OR status LIKE $5) AND
+				  updated_at > $6
+		);`
```

Comment:
Was just checking indexes for uploads table.

1. Do we keep deleting old entries from UploadsTable else without proper index on time columns this query can keep getting slower with time.
2. Not relevant to this PR. `CREATE INDEX wh_uploads_status_index ON wh_uploads(status text_ops);` index on status column with very low cardinality could have more performance drawbacks.

Reply:
> Do we keep deleting old entries from UploadsTable else without proper index on time columns this query can keep getting slower with time.

Yes. https://github.com/rudderlabs/rudder-server/blob/b51605af528cda22c1158f4daeae2675d92ee50a/warehouse/archive/archiver.go#L510

Reply:
> Not relevant to this PR. CREATE INDEX wh_uploads_status_index ON wh_uploads(status text_ops); index on status column with very low cardinality could have more performance drawbacks.

True, For columns with low cardinality, PostgreSQL might opt for a sequential scan over using the index, as scanning the table can be more efficient than traversing an index with low selectivity.

Reply:
Tracking Query-related things in [this](https://www.notion.so/rudderstacks/Warehouse-Query-Performance-175f2b415dd080d5bca6f78b1dd34c64?pvs=4) doc.